### ISSUE TITLE
Add a long name for the layers

### ISSUE THREAD

**Issue by charlesmartin14** [CalculatedContent/WeightWatcher#152]
## Currently, he way the layer name is set now, it only extracts the name of the kind of layer:  Embedding, Linear, Conv2D, etc
but it doesn't have a LongName

### if we look at a model like  **albert-large**  (with the MLM head for measuring its perplexity.)
the pytorch model gives me a of layer keys which can easily be understood: 

**keys:** 
['albert.embeddings.position_ids',
 'albert.embeddings.word_embeddings.weight', 
'albert.embeddings.position_embeddings.weight', 
'albert.embeddings.token_type_embeddings.weight', 
'albert.embeddings.LayerNorm.weight',
 'albert.embeddings.LayerNorm.bias', 
'albert.encoder.embedding_hidden_mapping_in.weight', 
'albert.encoder.embedding_hidden_mapping_in.bias', 'albert.encoder.albert_layer_groups.0.albert_layers.0.full_layer_layer_norm.weight', 
...
]

### This way weightwatcher works now, the weightwatcher walks the treee and detects the the leaf nodes, either by type or name, but it does not list the full name

<pre>
AlbertModel(
  (embeddings): AlbertEmbeddings(
    (word_embeddings): Embedding(30000, 128, padding_idx=0)
    (position_embeddings): Embedding(512, 128)
    (token_type_embeddings): Embedding(2, 128)
    (LayerNorm): LayerNorm((128,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0, inplace=False)
  )
  (encoder): AlbertTransformer(
    (embedding_hidden_mapping_in): Linear(in_features=128, out_features=768, bias=True)
    (albert_layer_groups): ModuleList(
      (0): AlbertLayerGroup(
        (albert_layers): ModuleList(
          (0): AlbertLayer(
            (full_layer_layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (attention): AlbertAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (attention_dropout): Dropout(p=0, inplace=False)
              (output_dropout): Dropout(p=0, inplace=False)
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            )
            (ffn): Linear(in_features=768, out_features=3072, bias=True)
            (ffn_output): Linear(in_features=3072, out_features=768, bias=True)
            (activation): NewGELUActivation()
            (dropout): Dropout(p=0, inplace=False)
          )
        )
      )
    )
  )
  (pooler): Linear(in_features=768, out_features=768, bias=True)
  (pooler_activation): Tanh()
)
</pre>

#### and weightwatcher sees each node in the tree as

<pre>
...
layer 14:AlbertAttention(
  (query): Linear(in_features=768, out_features=768, bias=True)
  (key): Linear(in_features=768, out_features=768, bias=True)
  (value): Linear(in_features=768, out_features=768, bias=True)
  (attention_dropout): Dropout(p=0, inplace=False)
  (output_dropout): Dropout(p=0, inplace=False)
  (dense): Linear(in_features=768, out_features=768, bias=True)
  (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
)
layer 15:Linear(in_features=768, out_features=768, bias=True)
layer 16:Linear(in_features=768, out_features=768, bias=True)
layer 17:Linear(in_features=768, out_features=768, bias=True)
layer 18:Dropout(p=0, inplace=False)
layer 19:Dropout(p=0, inplace=False)
layer 20:Linear(in_features=768, out_features=768, bias=True)
layer 21:LayerNorm((768,), eps=1e-12, elementwise_affine=True)
layer 22:Linear(in_features=768, out_features=3072, bias=True)
layer 23:Linear(in_features=3072, out_features=768, bias=True)
layer 24:NewGELUActivation()
...
</pre>

If we do extract a long name,  what would be the most desirable and useful thing ?

**Comment by quantitative-technologies** [comment]
Here is an example of what I am looking for. The model for my example is `albert-large` but with an SC head, so almost the same.

First I did `ww.analyze(albert_sc, min_evals=1)` so that layers would be kept even if they did not have enough eignevalues. That give the output:

![df-orig_2022-12-15_21-31-56](https://user-images.githubusercontent.com/29150871/208009870-5be29995-3563-47c6-85d3-618a7dd75bc6.png)

The full layer names are the keys you show above, which are too long add to the output (see below). 

I match a subset of the keys to the layers in the output by observing that WW only uses the `.weight` layers and the `norm` layers are not relevant:

```python
keys = list(albert_sc.state_dict().keys())
weight_idx = [i for i, k in enumerate(keys) if 'weight' in k]
layer_norm_idx = [i for i, k in enumerate(keys) if 'LayerNorm' in k]
norm_idx = [i for i, k in enumerate(keys) if 'norm' in k]
layer_idx = [i for i in weight_idx if i not in layer_norm_idx and i not in norm_idx]
```

Then I used a crude rule to extract a readable name from the key, described in the comments:

```python
def extract_layer_name(full_name):
  # truncate ".weight"
  k = full_name[:-7]
  # if there are at least 2 '.' symbols, then everything after the 2nd last one
  match = re.search("[^\.]+\.[^\.]+$", k)
  if match is not None:
    return match.group(0)
  # else all of k which has 1 or 2 words
  return k
```

It is not perfect, but now adding a `layer_name` column to the output I can clearly understand which layer of albert each row is describing:

![df_2022-12-15_21-33-41](https://user-images.githubusercontent.com/29150871/208011127-e5710072-aeba-425a-9925-4c2211233285.png)

**Comment by charlesmartin14** [comment]
The next release will contain a field, `longname`, which is the best full name, as indicated above

<img width="1112" alt="Screen Shot 2022-12-21 at 4 56 47 PM" src="https://user-images.githubusercontent.com/498448/209031779-fbea32b9-27c4-47b2-bb65-5259a4136d8b.png">

**Comment by charlesmartin14** [comment]
T
